packages/uni_app/lib/controller/parsers/schedule/new_api/parser.dart

```diff
@@ -0,0 +1,53 @@
+import 'dart:convert';
+
+import 'package:html/parser.dart' show parse;
+import 'package:http/http.dart' as http;
+import 'package:uni/controller/parsers/schedule/new_api/models/response_lecture.dart';
+import 'package:uni/model/entities/lecture.dart';
+
+/// Extracts the user's lecture API URL.
+///
+/// This function parses the schedule's HTML page.
+String getScheduleApiUrlFromHtml(
+  http.Response response,
+) {
+  final document = parse(response.body);
+
+  final scheduleElement = document.querySelector('#cal-shadow-container');
+  final apiUrl = scheduleElement?.attributes['data-evt-source-url'];
+
+  if (apiUrl == null) {
+    throw Exception('Could not find schedule API URL in schedule page');
+  }
+
+  return apiUrl;
+}
+
+List<Lecture> getLecturesFromApiResponse(
+  http.Response response,
+) {
+  final json = jsonDecode(response.body) as Map<String, dynamic>;
+  final data = json['data'] as List<dynamic>;
+
+  final now = DateTime.now();
+
+  return data
+      .cast<Map<String, dynamic>>()
+      .map(ResponseLecture.fromJson)
+      .where((lecture) => lecture.end.isAfter(now))
```

Comment:
From what I understood, this line filters the lectures to be only the ones that are not yet finished. However, the schedule page needs information about previous lectures that have already ended. Shouldn't this logic be only applied to the card?

Reply:
That is a good question and I actually asked it on Slack. You and @thePeras told me to do this :laughing:, but I can remove it if you want!